c++/src/kj/async-io-unix.c++

```diff
@@ -228,17 +239,77 @@ public:
     return promise.attach(kj::mv(fds), kj::mv(streams));
   }
 
+  Maybe<Promise<uint64_t>> tryPumpFrom(
+      AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+#if __linux__ && !__ANDROID__
+    KJ_IF_MAYBE(sock, kj::dynamicDowncastIfAvailable<AsyncStreamFd>(input)) {
+      return pumpFromOther(*sock, amount);
+    }
+#endif
+
+#if __linux__
+    KJ_IF_MAYBE(file, kj::dynamicDowncastIfAvailable<FileInputStream>(input)) {
+      KJ_IF_MAYBE(fd, file->getUnderlyingFile().getFd()) {
+        return pumpFromFile(*file, *fd, amount, 0);
+      }
+    }
+#endif
+
+    return nullptr;
+  }
+
+#if __linux__
+  // TODO(someday): Support sendfile on other OS's... unfortunately, it works differently on
+  //   different systems.
+
+private:
+  Promise<uint64_t> pumpFromFile(FileInputStream& input, int fileFd,
+                                 uint64_t amount, uint64_t soFar) {
+    while (soFar < amount) {
+      off_t offset = input.getOffset();
+      ssize_t n;
+
+      // Although sendfile()'s last argument has type size_t, on Linux it seems to cause EINVAL
+      // if we pass an amount that is greater than UINT32_MAX, so make sure to clamp to that. In
+      // practice, of course, we'll be limited to the socket buffer size.
+      size_t requested = kj::min(amount - soFar, (uint32_t)kj::maxValue);
+
+      KJ_SYSCALL_HANDLE_ERRORS(n = sendfile(fd, fileFd, &offset, requested)) {
+        case EINVAL:
+        case ENOSYS:
+          // Fall back to regular pump
+          return unoptimizedPumpTo(input, *this, amount, soFar);
+
+        case EAGAIN:
+          return observer.whenBecomesWritable()
+              .then([this, &input, fileFd, amount, soFar]() {
+            return pumpFromFile(input, fileFd, amount, soFar);
```

Comment:
As absolutely unlikely as it will be to get `EAGAIN` multiple times in a row, the possibility is still non-zero. Something to guard against too much recursion might be good. But also happy like this.

Reply:
From the man page:

`EAGAIN Nonblocking I/O has been selected using O_NONBLOCK and the write would block.`

`EAGAIN` here very specifically means that the socket is not writable because the send buffer is full. So we wait until it becomes writable, and try again. Since we're using edge triggering, it is technically possible that we could receive a queued writability edge notification that is no longer accurate, but we would not receive a never-ending stream of these, barring some sort of [deep kernel bug](https://sandstorm.io/news/2015-04-08-osx-security-bug).

Note also that this is promise recursion, and it's a tail call, so this does not result in a growing chain.

Reply:
Yep, I'm just in an overly paranoid mood at the moment ;-) ... as I said, happy with this as is.